console/src/state/resources.rs

```diff
@@ -0,0 +1,332 @@
+    resources: HashMap<u64, Rc<RefCell<Resource>>>,
+    new_resources: Vec<ResourceRef>,
+}
+
+#[derive(Debug, Copy, Clone)]
+#[repr(usize)]
+pub(crate) enum SortBy {
+    Rid = 0,
+    Kind = 1,
+    ConcreteType = 2,
+    Target = 3,
+    Total = 4,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub(crate) enum Kind {
+    Timer,
+    Other(String),
+}
+
+#[derive(Debug)]
+pub(crate) struct Resource {
+    id: u64,
+    meta_id: u64,
+    kind: Kind,
+    stats: ResourceStats,
+    target: Arc<str>,
+    concrete_type: String,
+    location: String,
+}
+
+pub(crate) type ResourceRef = Weak<RefCell<Resource>>;
+
+#[derive(Debug)]
+struct ResourceStats {
+    created_at: SystemTime,
+    dropped_at: Option<SystemTime>,
+    total: Option<Duration>,
+    attributes: Vec<Attribute>,
+    formatted_attributes: Vec<Vec<Span<'static>>>,
+}
+
+#[derive(Debug)]
+pub(crate) struct Attribute {
+    field: Field,
+    unit: Option<String>,
+}
+
+impl Default for SortBy {
+    fn default() -> Self {
+        Self::Rid
+    }
+}
+
+impl SortBy {
+    pub fn sort(&self, now: SystemTime, resources: &mut Vec<Weak<RefCell<Resource>>>) {
+        match self {
+            Self::Rid => {
+                resources.sort_unstable_by_key(|resource| resource.upgrade().map(|r| r.borrow().id))
+            }
+            Self::Kind => resources.sort_unstable_by_key(|resource| {
+                resource.upgrade().map(|r| r.borrow().kind.clone())
+            }),
+            Self::ConcreteType => resources.sort_unstable_by_key(|resource| {
+                resource.upgrade().map(|r| r.borrow().concrete_type.clone())
+            }),
+            Self::Target => resources.sort_unstable_by_key(|resource| {
+                resource.upgrade().map(|r| r.borrow().target.clone())
+            }),
+            Self::Total => resources
+                .sort_unstable_by_key(|resource| resource.upgrade().map(|r| r.borrow().total(now))),
+        }
+    }
+}
+
+impl TryFrom<usize> for SortBy {
+    type Error = ();
+    fn try_from(idx: usize) -> Result<Self, Self::Error> {
+        match idx {
+            idx if idx == Self::Rid as usize => Ok(Self::Rid),
+            idx if idx == Self::Kind as usize => Ok(Self::Kind),
+            idx if idx == Self::ConcreteType as usize => Ok(Self::ConcreteType),
+            idx if idx == Self::Target as usize => Ok(Self::Target),
+            idx if idx == Self::Total as usize => Ok(Self::Total),
+            _ => Err(()),
+        }
+    }
+}
+
+impl view::SortBy for SortBy {
+    fn as_column(&self) -> usize {
+        *self as usize
+    }
+}
+
+impl ResourcesState {
+    pub(crate) fn take_new_resources(&mut self) -> impl Iterator<Item = ResourceRef> + '_ {
+        self.new_resources.drain(..)
+    }
+
+    pub(crate) fn update_resources(
+        &mut self,
+        styles: &view::Styles,
+        update: proto::resources::ResourceUpdate,
+        metas: &HashMap<u64, Metadata>,
+        clear_new_list: bool,
```

Comment:
I find that functions with boolean arguments are less clear, especially when the function becomes more complex (Simple yes-no functions are fine). I'd personally advocate for a more specific enum argument, even if it's just this "yes" and "no" variant names. I think in this case we can make a nicer one:

```rust
enum Visibility {
    Show,
    Hide,
}
```